### PR TITLE
[FLINK-19496][table] DataGen source DECIMAL always returns null

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

DataGen source DECIMAL always returns null

## Brief change log

Use longs as the base generator to avoid precision issues. 

## Verifying this change

UT

## Does this pull request potentially affect one of the following parts:
No

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Review by knaufk — CHANGES_REQUESTED** [review]
I think, we should properly fix this in the sense that a the datagen source produces DECIMAL values with a precision and scale as specified in the schema, e.g. DECIMAL(6,3) produces values like: 

129.231
234.523
...

The current solution again produces "null" for e.g. DECIMAL(6,3).

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	private final int scale;
+
+	private final double min;
+
+	private final double max;
+
+	private final String fmt;
+
+	public DecimalDataRandomGenerator(int precision, int scale, double min, double max) {
+		double largest = largestValue(precision, scale);
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:47]:
```suggestion
		double largest = Math.pow(10,precision-scale) - Math.pow(10,-scale);
```

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		for (int i = 0; i < precision; i++) {
+			if (scale != 0 && precision != scale && i + scale == precision) {
+				builder.append(".");
+			}
+
+			builder.append(9);
+		}
+
+		return Double.parseDouble(builder.toString());
+	}
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:90]:
```suggestion
```

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	@Override
+	public boolean hasNext() {
+		return true;
+	}
+
+	@Override
+	public DecimalData next() {
+		String value = String.format(fmt, ThreadLocalRandom.current().nextDouble(min, max));
+
+		BigDecimal decimal = new BigDecimal(value.trim(), new MathContext(precision, RoundingMode.DOWN));
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:68]:
```suggestion
		BigDecimal decimal = new BigDecimal(ThreadLocalRandom.current().nextDouble(min, max), new MathContext(precision, RoundingMode.DOWN));
```

Comment (knaufk) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:68]: Doesn't this also work like that?

File: flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Tests that the data generator is valid for every combination of precision and scale.
+ */
+public class DecimalDataRandomGeneratorTest {
+
+	@Test
+	public void testGeneratesNonNullValues() {
+		for (int precision = 1; precision <= 38; precision++) {
+			for (int scale = 1; scale <= precision; scale++) {
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/test/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGeneratorTest.java:37]:
```suggestion
			for (int scale = 0; scale <= precision; scale++) {
```

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+	private final String fmt;
+
+	public DecimalDataRandomGenerator(int precision, int scale, double min, double max) {
+		double largest = largestValue(precision, scale);
+		this.precision = precision;
+		this.scale = scale;
+		this.min = Math.min(-1 * largest, min);
+		this.max = Math.min(largest, max);
+		this.fmt = "%" + precision + "." + scale + "f";
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:52]:
```suggestion
```

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	private final int precision;
+
+	private final int scale;
+
+	private final double min;
+
+	private final double max;
+
+	private final String fmt;
```
Comment (knaufk) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:44]:
```suggestion
```

**Review by knaufk — CHANGES_REQUESTED** [review]
Thanks for addressing my comments. I left some suggestions for simplifications. 

Generally, I think, the generation based on a random double is a bit flawed (using a random BigInteger would probably be cleaner), but good enough.

Could you add test that sets a min/max value and asserts that the result is within the bounds?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java (hunk 3 above)
Comment (sjwiesman) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/datagen/types/DecimalDataRandomGenerator.java:68]: yes

**Comment by sjwiesman** [comment]
@knaufk thanks for taking a look, I've pushed a fix that should address all of your comments.

**Review by knaufk — CHANGES_REQUESTED** [review]
Thanks for the additional tests. Fix Checkstyle and we are good to go.

**Review by knaufk — APPROVED** [review]
Thanks. "mvn clean install" runs locally for flink-table-api-java-bridge for me now. Merging.

**Pull request merged**